### ISSUE TITLE
Saving checkpoint stops indefinitely when using LoRA under DDP

### ISSUE THREAD

**Issue by timothylimyl** [Lightning-AI/lit-llama#252]
Hi there,

I tried running `finetune_lora.py` with DDP (4 devices) and the scripts always run fine up until the 3rd time it has to save the checkpoint. It just hangs at device 1 saving checkpoint. The bug is weird to me cause it manages to save for the first 2 times (all devices prints and saves accordingly as per: `fabric.save(os.path.join(out_dir, f"iter-{iter_num}-loss-{val_loss:.4f}-ckpt.pth"), checkpoint)`

Is this some bug with fabric?

**Comment by timothylimyl** [comment]
Just to confirm, I tried running it a few times and can confirm that it hangs randomly (not 3rd time of saving checkpoint). For example, my last run, it hanged after saving around 8 checkpoints at GPU No.3:
```
Saving LoRA weights to out/lora/all_custom_run2
Saving LoRA weights to out/lora/all_custom_run2
Saving LoRA weights to out/lora/all_custom_run2

```
Usually it does hang at the first GPU though so it just prints one `Saving LoRA weights to out/lora/all_custom_run2` and then stops there indefinitely (I tried waiting a few hours).

**Comment by awaelchli** [comment]
Just a random guess, maybe you need to insert a `fabric.barrier()` call right after `fabric.save()`. This is automatically done already in the Fabric version on Lightning master, but for 2.0.0 it is probably required. 

> Edit 1: Could it be that we need to only save when global rank==0?

No, this is handled by Fabric.

**Comment by timothylimyl** [comment]
I am on the master branch version already, same issue. It hangs without any error popping up, then I am forced to switch off the terminals to shut it down.
